File: src/main/java/app/application/data/entities/YoutubePlaylist.java
package app.application.data.entities;

import com.github.kiulian.downloader.model.playlist.PlaylistInfo;
import com.github.kiulian.downloader.model.playlist.PlaylistVideoDetails;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import lombok.Getter;

import java.util.ArrayList;
import java.util.List;

public class YoutubePlaylist extends YoutubeEntity<PlaylistInfo> {

	private final List<YoutubePlaylistVideoDetail> youtubePlaylistVideoDetailsCache = new ArrayList<>();

	@Getter
	private final BooleanProperty audioOnly;

	public YoutubePlaylist(PlaylistInfo playlistInfo){
		this.reference = playlistInfo;
		this.audioOnly = new SimpleBooleanProperty(false);
	}

	public int getPlaylistSize(){
		return reference.details().videoCount();
	}

	public List<YoutubePlaylistVideoDetail> getPlaylistVideos(){
		if(youtubePlaylistVideoDetailsCache.isEmpty()){
			for (PlaylistVideoDetails video : reference.videos()) {
				youtubePlaylistVideoDetailsCache.add(new YoutubePlaylistVideoDetail(video));
			}
		}
		return youtubePlaylistVideoDetailsCache;
	}

	public String getPlaylistTitle() {
		return reference.details().title();
	}


}

File: src/main/kotlin/app/application/data/entities/YoutubePlaylist.kt
package app.application.data.entities

import com.github.kiulian.downloader.model.playlist.PlaylistInfo
import javafx.beans.property.BooleanProperty
import javafx.beans.property.SimpleBooleanProperty

class YoutubePlaylist(playlistInfo: PlaylistInfo?) : YoutubeEntity<PlaylistInfo?>() {
    private val youtubePlaylistVideoDetailsCache: MutableList<YoutubePlaylistVideoDetail> = ArrayList()

    var audioOnly: BooleanProperty

    init {
        reference = playlistInfo
        audioOnly = SimpleBooleanProperty(false)
    }

    val playlistSize: Int
        get() = reference?.details()?.videoCount() ?: 0

    val playlistVideos: List<YoutubePlaylistVideoDetail>
        get() {
            if (youtubePlaylistVideoDetailsCache.isEmpty()) {
                reference?.also {
                    for (video in it.videos()) {
                        youtubePlaylistVideoDetailsCache.add(YoutubePlaylistVideoDetail(video))
                    }
                }
            }
            return youtubePlaylistVideoDetailsCache
        }

    val playlistTitle: String
        get() = reference?.details()?.title() ?: "No Playlist"
}